### DISCUSSION TITLE
Failed to read parameter

### DISCUSSION THREAD

**Discussion by TypeFloat** [bulletphysics/bullet3#4139] [Q&A]
When I use "addUserDebugParameter", I try to use "readUserDebugParameter" to get the realtime parameter, but I get an error about "failed to read paramter". What makes me crazy is that sometimes it runs successfully while sometimes I catch the error. Could you help me?

**Comment by TypeFloat** [comment]
I find that when I use time.sleep(1/60) it runs successfully. What happened? I can't call the read... too frequently?

**Comment by rohit-kumar-j** [comment]
@TypeFloat, need the example to reproduce the error. I've not had an error with `addUserDebugParameter` or `readUserDebugParameter` yet.

**Comment by tw-ilson** [comment]
I have been having the same issue. I think it's only with buttons? This example produces the issue for me:

      def read_parameters(dbg_params):
      
          values = dict()
          for name, param in dbg_params.items():
              try:
                  values[name] = pb.readUserDebugParameter(param)
              except:
                  traceback.print_exc()
                  print(f"PARAM {param}")
      
          return values
      
      def interactive_transformations(axis_id):
          np.set_printoptions(suppress=True, precision=4)
      
          dbg = dict()
          # for view matrix
          translation_params = ['x', 'y', 'z']
          for p in translation_params:
              dbg[p] = pb.addUserDebugParameter(p, -10, 10, 0)
      
          rotation_params = ['roll', 'pitch', 'yaw']
          for p in rotation_params:
              dbg[p] = pb.addUserDebugParameter(p, -180, 180, 0)
      
          dbg['print'] =  pb.addUserDebugParameter('print details', 1, 0, 1)
          old_print_val = 1
          while 1:
              values = read_parameters(dbg)
      
              pos = np.array([values[c] for c in translation_params])
              euler = np.radians([values[c] for c in rotation_params])
      
              #update visual aids for camera, target pos
              quat = np.array(pb.getQuaternionFromEuler(euler))
              pb.resetBasePositionAndOrientation(axis_id, pos, quat)
      
              T_mtx = np.eye(4)
              T_mtx[:3,:3] = R.from_quat(quat).as_matrix()
              T_mtx[:3,3] = pos
      
              if old_print_val != values['print']:
                  old_print_val = values['print']
                  print("\n========================================")
                  print(f"translation = {pos}")
                  print(f"euler angles = {euler}")
                  print(f"quat = {quat}")
                  T_mtx_str = f"{T_mtx}".splitlines()
                  print(f"T_mtx = {T_mtx_str[0]}")
                  print(f"        {T_mtx_str[1]}")
                  print(f"        {T_mtx_str[2]}")
                  print(f"        {T_mtx_str[3]}")
                  print("========================================\n")
      
      if __name__ == "__main__":
          pb.connect(pb.GUI)
          #pb.configureDebugVisualizer(pb.COV_ENABLE_GUI, 0)
      
          # add axis
          axis_id = pb.loadURDF('assets/urdf/axis.urdf', globalScaling=0.5)
      
          interactive_transformations(axis_id)

**Comment by rohit-kumar-j** [comment]
@tw-ilson, Perhaps you mean this:
([pybullet quickstart guide](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#))
![image](https://user-images.githubusercontent.com/37873142/151032878-4d00a168-055f-4a71-a27d-bf001a1c14a8.png)
![image](https://user-images.githubusercontent.com/37873142/151033051-dbf87f51-6a32-4257-8da3-11422b6778d8.png)

**Reply by tw-ilson** [reply]
Sorry, to be clearer, the program output is the following. The program is crashing as the serialized parameters are being created in the GUI, before any interaction. This happens in DIRECT mode as well.  Were you able to reproduce this error? @TypeFloat 's solution to sleep the program works for me, but the program still crashes after a time. 

```
Version = 4.1 Metal - 71.7.1
Vendor = Apple
Renderer = Apple M1
b3Printf: Selected demo: Physics Server
startThreads creating 1 threads.
starting thread 0
started thread 0 
MotionThreadFunc thread started
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: No inertial data for link, using mass=1, localinertiadiagonal = 1,1,1, identity local inertial frame
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: x_vec
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: No inertial data for link, using mass=1, localinertiadiagonal = 1,1,1, identity local inertial frame
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: y_vec
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: No inertial data for link, using mass=1, localinertiadiagonal = 1,1,1, identity local inertial frame
b3Printf: b3Warning[examples/Importers/ImportURDFDemo/BulletUrdfImporter.cpp,126]:

b3Printf: z_vec
Traceback (most recent call last):
  File "rotation_translation_visualizer.py", line 25, in read_parameters
    values[name] = pb.readUserDebugParameter(param)
pybullet.error: Failed to read parameter.
PARAM 0
Traceback (most recent call last):
  File "rotation_translation_visualizer.py", line 82, in <module>
    interactive_transformations(axis_id)
  File "rotation_translation_visualizer.py", line 51, in interactive_transformations
    pos = np.array([values[c] for c in translation_params])
  File "rotation_translation_visualizer.py", line 51, in <listcomp>
    pos = np.array([values[c] for c in translation_params])
KeyError: 'x'
numActiveThreads = 0
stopping threads
Thread with taskId 0 exiting
Thread TERMINATED
destroy semaphore
semaphore destroyed
destroy main semaphore
main semaphore destroyed
```

**Comment by TypeFloat** [comment]
Do you think the reason is that we all use apple m1 silicon? The solution I proposed had the same problem, as I later discovered. @tw-ilson

**Reply by tw-ilson** [reply]
It is possible, though I know people on the same system that have not had the issue (as far as I know).
